scripts/lib/CIME/XML/generic_xml.py

```diff
@@ -91,7 +104,7 @@ def read(self, infile, schema=None):
 
             logger.debug("File version is {}".format(str(self.get_version())))
 
-            self._FILEMAP[infile] = (self.tree, self.root)
+            self._FILEMAP[infile] = (self.tree, self.root, os.path.getmtime(infile))
```

Comment:
It seems like there's a possible race condition due to the timing of when you call getmtime to set the time of the cached data. Consider this scenario:

- Starting point: the mtime of the given file is 100.

- The current process needs to read from file. It opens the file and reads it.

- Meanwhile, another process writes to this file, resulting in an updated mtime (say, 110).

- Then the current process calls getmtime, getting the value 110. At this point, the current process would think it has data from mtime=110, whereas in fact it just has the data that were current as of mtime=100.

- Later, another read is done (without any more writes happening). It would use the cached data (because it incorrectly thinks it has the data from mtime=110), whereas in fact it should reread the file.

I think that moving the call to getmtime to _before_ the file_open would fix this problem. You could have the opposite problem – the current process thinks the data are out-of-date when in fact they're up-to-date – but that problem is less severe (just leading to a loss of efficiency, not incorrect data).

Reply:
@billsacks you are exactly right and this is the point I was trying to make in the discussion ticket that we will never achieve 100% robustness unless we access entry values through some kind of database system.

The reason my approach works (I think) is that I've made the simplifying assumption that CIME will never be used in the way you described. We've already discussed that the bulk of CIME is not thread safe, there's far too much shared state. Due to the situation with the XML files, it is also not safe to run multiple CIME processes in the same case concurrently. Note that all the CIME subprocess calls are synchronous calls (the parent spawns the child and then waits for it to finish before doing anything else), so no true concurrency is happening. The subprocesses can modify XML and, as long as the parent refreshes the cache when the subprocess is done, everything is OK.

Reply:
Okay, fair enough - thanks for the explanation.